### PR TITLE
Enable tests against existing workload cluster

### PR DESCRIPTION
**Pull request by jsturtevant**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
When developing new e2e test specs or working on the framework it is time consuming to wait for a cluster to be provisioned.  This enables reusing existing workload cluster:

Assuming you already have a kind management cluster and ran e2e once which created a workload cluster called `capz-e2e-uf9yll`:

```
export LOCAL_ONLY=true 
export SKIP_CLEANUP=true 
export GINKGO_FOCUS=".*Windows.Enabled.*" 
export WINDOWS=true
export CLUSTERNAME=capz-e2e-uf9yll
export SKIP_CREATE_MGMT_CLUSTER=true
export KUBECONFIG=/home/jstur/projects/cluster-api-provider-azure/kubeconfig.e2e

./scripts/ci-e2e.sh
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [x] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: docs/development.md [hunk 1]
```diff
@@ -403,6 +403,7 @@ You can optionally set the following variables:
 | `SKIP_CREATE_MGMT_CLUSTER` | Skip management cluster creation. If skipping managment cluster creation you must specify `KUBECONFIG` and `SKIP_CLEANUP` | `false`                                                                               |
 | `LOCAL_ONLY`               | Use Kind local registry and run the subset of tests which don't require a remotely pushed controller image.               | `true`                                                                                |
 | `REGISTRY`                 | Registry to push the controller image.                                                                                    | `capzci.azurecr.io/ci-e2e`                                                            |
+| `CLUSTERNAME`              | Name of an exisiting workload cluster.  Will run specs against existing workloadcluster. Use in conjunction with `SKIP_CREATE_MGMT_CLUSTER` and `KUBECONFIG` | `capz-e2e-uf9yll`                                       |
```
Comment (chewong) [inline docs/development.md:406]:
```suggestion
| `CLUSTER_NAME`              | Name of an exisiting workload cluster.  Will run specs against existing workload cluster. Use in conjunction with `SKIP_CREATE_MGMT_CLUSTER` and `KUBECONFIG` | `capz-e2e-uf9yll`                                       |
```

File: test/e2e/common.go [hunk 2]
```diff
@@ -61,13 +62,31 @@ func Byf(format string, a ...interface{}) {
 	By(fmt.Sprintf(format, a...))
 }
 
-func setupSpecNamespace(ctx context.Context, specName string, clusterProxy framework.ClusterProxy, artifactFolder string) (*corev1.Namespace, context.CancelFunc) {
-	Byf("Creating a namespace for hosting the %q test spec", specName)
+func setupSpecNamespace(ctx context.Context, namespaceName string, clusterProxy framework.ClusterProxy, artifactFolder string) (*corev1.Namespace, context.CancelFunc) {
+	Byf("Creating a namespace for hosting the %q test spec", namespaceName)
+	logpath := filepath.Join(artifactFolder, "clusters", clusterProxy.GetName())
```
Comment (chewong) [inline test/e2e/common.go:67]:
```suggestion
	logPath := filepath.Join(artifactFolder, "clusters", clusterProxy.GetName())
```

File: test/e2e/common.go [hunk 3]
```diff
@@ -61,13 +62,31 @@ func Byf(format string, a ...interface{}) {
 	By(fmt.Sprintf(format, a...))
 }
 
-func setupSpecNamespace(ctx context.Context, specName string, clusterProxy framework.ClusterProxy, artifactFolder string) (*corev1.Namespace, context.CancelFunc) {
-	Byf("Creating a namespace for hosting the %q test spec", specName)
+func setupSpecNamespace(ctx context.Context, namespaceName string, clusterProxy framework.ClusterProxy, artifactFolder string) (*corev1.Namespace, context.CancelFunc) {
+	Byf("Creating a namespace for hosting the %q test spec", namespaceName)
```
Comment (chewong) [inline test/e2e/common.go:66]: `namespaceName` is not the test spec name so maybe we can add `specName` as part of the function signature

File: test/e2e/common.go [hunk 4]
```diff
@@ -61,13 +62,31 @@ func Byf(format string, a ...interface{}) {
...
+			defer GinkgoRecover()
+			framework.WatchNamespaceEvents(watchesCtx, framework.WatchNamespaceEventsInput{
+				ClientSet: clusterProxy.GetClientSet(),
+				Name:      namespace.Name,
+				LogFolder: logpath,
+			})
+		}()
+
+		return namespace, cancelWatches
+	}
```
Comment (chewong) [inline test/e2e/common.go:83]: In case of an error, we should exit unless it's an API server not found error

File: docs/development.md (hunk 1 above)
Comment (jsturtevant) [reply docs/development.md:406]: will also need to update the env variable

File: test/e2e/common.go (hunk 3 above)
Comment (jsturtevant) [reply test/e2e/common.go:66]: The speName bring much here, how about rewording to `Byf("Creating namespace %q for hosting the cluster", namespaceName)`

File: test/e2e/common.go (hunk 4 above)
Comment (jsturtevant) [reply test/e2e/common.go:83]: good catch

File: docs/development.md [hunk 5]
```diff
@@ -403,6 +403,7 @@ You can optionally set the following variables:
 | `SKIP_CREATE_MGMT_CLUSTER` | Skip management cluster creation. If skipping managment cluster creation you must specify `KUBECONFIG` and `SKIP_CLEANUP` | `false`                                                                               |
 | `LOCAL_ONLY`               | Use Kind local registry and run the subset of tests which don't require a remotely pushed controller image.               | `true`                                                                                |
 | `REGISTRY`                 | Registry to push the controller image.                                                                                    | `capzci.azurecr.io/ci-e2e`                                                            |
+| `CLUSTER_NAME`              | Name of an exisiting workload cluster.  Will run specs against existing workload cluster. Use in conjunction with `SKIP_CREATE_MGMT_CLUSTER` and `KUBECONFIG` | `capz-e2e-uf9yll`                                       |
```
Comment (CecileRobertMichon) [inline docs/development.md:406]: one test run != one workload cluster, this might lead to unexpected results if not used with GINKGO_FOCUS (assuming it will re-apply the template for each spec using the same cluster names, especially if running the different specs in parallel)

Comment (jsturtevant) [reply docs/development.md:406]: Yes. I will update the docs to point that out.

Comment (jsturtevant) [reply docs/development.md]: updated to clarify

**Comment by jsturtevant** [comment]
the windows failure was 

```
Delete "https://capz-e2e-7xp181-620890e5.westus2.cloudapp.azure.com:6443/api/v1/namespaces/default/services/webqki40r-elb": http2: client connection lost
```

Haven't seen this error before. will keep an eye out for it
/test pull-cluster-api-provider-azure-e2e-windows

File: docs/development.md [hunk 6]
```diff
@@ -403,6 +403,7 @@ You can optionally set the following variables:
 | `SKIP_CREATE_MGMT_CLUSTER` | Skip management cluster creation. If skipping managment cluster creation you must specify `KUBECONFIG` and `SKIP_CLEANUP` | `false`                                                                               |
 | `LOCAL_ONLY`               | Use Kind local registry and run the subset of tests which don't require a remotely pushed controller image.               | `true`                                                                                |
 | `REGISTRY`                 | Registry to push the controller image.                                                                                    | `capzci.azurecr.io/ci-e2e`                                                            |
+| `CLUSTER_NAME`             | Name of an exisiting workload cluster.  Will run specs against existing workload cluster. Use in conjunction with `SKIP_CREATE_MGMT_CLUSTER`, `GINKGO_FOCUS` and `KUBECONFIG`. Must specify only one e2e spec to run against with `GINKGO_FOCUS` such as `export GINKO_FOCUS=Creating.a.VMSS.cluster.with.a.single.control.plane.node`. | `capz-e2e-uf9yll`                                       |
```
Comment (CecileRobertMichon) [inline docs/development.md:406]: what's `capz-e2e-uf9yll` ? there shouldn't be a default for this

File: test/e2e/azure_test.go [hunk 7]
```diff
@@ -58,10 +59,17 @@ var _ = Describe("Workload cluster creation", func() {
 		Expect(e2eConfig.Variables).To(HaveKey(capi_e2e.KubernetesVersion))
 		Expect(e2eConfig.Variables).To(HaveKey(capi_e2e.CNIPath))
 
+		clusterName = os.Getenv("CLUSTER_NAME")
+		if clusterName == "" {
+			clusterName = fmt.Sprintf("capz-e2e-%s", util.RandomString(6))
+		}
+		fmt.Fprintf(GinkgoWriter, "INFO: using cluster %s", clusterName)
```
Comment (CecileRobertMichon) [inline test/e2e/azure_test.go:66]: can we only include this log line when clusterName is not empty and was provided?

File: test/e2e/conformance_test.go [hunk 8]
```diff
@@ -56,10 +56,17 @@ var _ = Describe("Conformance Tests", func() {
 		Expect(e2eConfig.Variables).To(HaveKey(capi_e2e.KubernetesVersion))
 		Expect(e2eConfig.Variables).To(HaveKey(capi_e2e.CNIPath))
 
+		clusterName = os.Getenv("CLUSTER_NAME")
+		if clusterName == "" {
+			clusterName = fmt.Sprintf("capz-conf-%s", util.RandomString(6))
+		}
+		fmt.Fprintf(GinkgoWriter, "INFO: using cluster %s", clusterName)
```
Comment (CecileRobertMichon) [inline test/e2e/conformance_test.go:63]: ditto

File: test/e2e/azure_test.go (hunk 7 above)
Comment (jsturtevant) [reply test/e2e/azure_test.go:66]: I thought this was useful in both cases. If it is Randomly generated it is helpful to have that to be clearly specified in the out.  Maybe "using cluster" is the wrong wording here and should be something like "cluster name is"?

Comment (CecileRobertMichon) [reply test/e2e/azure_test.go:66]: Yeah, +1 to "Using cluster name" or "cluster name is"

File: docs/development.md (hunk 6 above)
Comment (jsturtevant) [reply docs/development.md:406]: will remove, thought it was examples but see now that it is defaults

Comment (jsturtevant) [reply docs/development.md]: removed

File: test/e2e/azure_test.go (hunk 7 above)
Comment (jsturtevant) [reply test/e2e/azure_test.go]: updated to `Cluster name is`

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve

**Comment by jsturtevant** [comment]
The failure here was #1328 which i've had some success fixing in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1351#issuecomment-830418597

/test pull-cluster-api-provider-azure-e2e-windows
